index.bs

```diff
@@ -989,6 +989,10 @@ interface MLGraphBuilder {
 };
 </script>
 
+<div class="note">
+The {{MLGraphBuilder}}.{{MLGraphBuilder/build()}} method compiles the graph builder state up to the specified output operands into a compiled graph according to the type of {{MLContext}} backing it. When the {{[[contextType]]}} of the {{MLContext}} is set to [=default-context|default=] with the {{MLContextOptions}}.{{deviceType}} set to [=device-type-gpu|gpu=], the compiled graph is initialized right before the {{MLGraphBuilder/build()}} method call returns. This initialization stage is important for optimal performance of the subsequent graph executions. See [[#api-mlcommandencoder-graph-initialization]] for more detail.
```

Comment:
This initialization stage should also be important for cpu device. According to the existing CPU backends of WebNN-native, including XNNPack, oneDNN or MLAS, all of them will re-layout/re-order the static weights to optimized memory layout (packed/blocked) for optimal performance.

As the `build` method is a sync call, it would block until the graph is initialized. It would impact the responsiveness of UI when calling it on the main thread.

Reply:
On responsiveness I think you can make the same argument even without considering weight processing. A straight-up compilation process could be expensive regardless. Perhaps, it should be constrained to a worker thread as well?

Reply:
You are right, actually I mentioned it in the original issue report https://github.com/webmachinelearning/webnn/issues/230.  

> Perhaps, it should be constrained to a worker thread as well?

Yes, we probably can do the same for the `build` as the `compute` that restricts the sync `build` method in the worker thread and introduces aync `buildAsync` method for the main thread.

Reply:
How about 0ea819d7d57b000cce5188aa613fe5571f8de463 now?

Reply:
How about this 0ea819d7d57b000cce5188aa613fe5571f8de463?

Reply:
Thanks @wchao1115 , https://github.com/webmachinelearning/webnn/commit/0ea819d7d57b000cce5188aa613fe5571f8de463 looks good to me.

Should we support asynchronous graph initialization by introducing `buildAsync` for main thread usage? Because without that, the only way to build an `MLGraph` is calling `build` in a worker. It means, for asynchronous execution in the main thread, the caller has to transfer the `MLGraph` back to the main thread. This may be inconvenient and inefficient. So I propose we define `buildAsync` like:

```js
Promise<MLGraph> buildAsync(MLNamedOperands outputs);
```

Any thoughts?

Reply:
Should we change `build()` to return `Promise<MLGraph>`? Having the API default to async gives us lots of flexibility in the future.

Also, it's easy to write async calls in synchronous style with JS async-await. Converting sync / blocking calls to async / non-blocking calls is much harder.

Reply:
Thanks @wacky6 

> Also, it's easy to write async calls in synchronous style with JS async-await.

It's true. 

However, it is hard or inefficient for a C++ inference code of Wasm-based ML framework to integrate. That C++ inference code usually expects a sync method call. The sync methods (build/compute) are designed for that use case.

My suggestion is to keep the sync `build` method but restrict it in worker scope and introduce the async `buildAsync` for main thread / JS usage.

Reply:
Thanks for the explanation!

I think there're libraries to wrap async Web API calls in JavaScript. Like: https://web.dev/asyncify/ .

Are there any constraints limiting such libraries wrapping async WebNN methods to WASM sync method calls?

For Web APIs in general, I don't think "Async" suffic to JavaScript API is used elsewhere. Async functions are first-class citizens in JS, and most web developers expect async APIs nowadays.

If `buildAsync()` can be trivially converted to `buildSync` in WASM, I think just providing the `Promise<MLGraph> build()` method is better in terms of simplicity and expandability.

Reply:
Thanks @wacky6 and @huningxin for the feedback. I've created issue #263 to track this request separately.

Reply:
Thanks @wchao1115 . I have no other comments for this PR.